### PR TITLE
support DTensor S(1) -> Partial("sum")

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

  Add DTensor support for Shard(dim) -> Partial("sum") redistribution
  using local zero-fill. This enables the S(1) -> P transition required
  by TorchTitan's new MoE sharding
  (https://github.com/pytorch/torchtitan/pull/3996).

  ## Testing

  Added forward, backward, uneven-sharding, and graph-planner coverage.
```
python test/distributed/tensor/test_redistribute.py -k shard_to_partial
```

### CODE REVIEW

**Comment by weifengpy** [comment]
I support the motivation. From DTensor perspective, is your intention to gate Shard-to-Partial to avoid changing exsting callsite? current this PR changes _StridedShard behavior. Repro as below

here is my prompt "code review code-reivew/pytorch 4356b41b248f8211b25226091dd82f5468f1357f . identify if it breaks existing callsite because the commit allowed Shard to Partial redistribution"

_StridedShard is the exercised path for FSDP2 + TP and FSDP2 + EP when they shard on the same dim

```
import os
import torch
import torch.distributed as dist
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.tensor import distribute_tensor, Partial, Shard
from torch.distributed.tensor.placement_types import _StridedShard
torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
dist.init_process_group("nccl")
mesh = init_device_mesh("cuda", (2, 2))
ss = _StridedShard(0, split_factor=2)
source = distribute_tensor(torch.arange(8, device="cuda"), mesh, (ss, Shard(0)))
result = source.redistribute(mesh, (ss, Partial("sum")))
```

**Comment by acisseJZhong** [comment]
> is your intention to gate Shard-to-Partial to avoid changing exsting callsite?

My intention was to avoid the planner exposes S -> P, the planer considers S -> P only when it is the final requested destination.

regarding to the breakage: instead of `(SS(0, split_factor=2), Shard(0))` ->  `(SS(0, split_factor=2), Partial("sum"))`, I think the correct behavior should be `(SS(0, split_factor=2), Shard(0))` -> `(Shard(0), Partial('sum'))`? 

However I checked that before this PR,  `(SS(0), S(0))` ->`(SS(0), R)` will raise error. 
https://github.com/acisseJZhong/pytorch/blob/d2f6cf4c3c76f6043523f442550d2c8e91fa1d9f/torch/distributed/tensor/_redistribute.py#L1410, we should also raise error if we ever have  `(SS(0), S(0))` ->`(SS(0), P)` rather than silently breaking numerics?

Updated the code with the added check `torch._check(local_tensor.size(self.dim) == local_shard_size)` during zero padding, so it will explicitly fail, similarly to how we fail with `(SS(0), S(0))` ->`(SS(0), R)`.

**Comment by wconstab** [comment]
Remembering the context.. 

X->Partial("sum") can be numerically dangerous if you do it by scaling, or can be SPMD-dangerous if you do it by padding (different ranks have different paddings, which means your graphs are non-spmd).

IIUC your PR is taking the latter approach, padding different sized chunks per rank as prefix/suffix.

I would prefer not to support this in the current form, as we want DTensor to produce SPMD code.  I am not familiar with the current MoE sharding situation though.  A few thoughts
* if we're going to lift the ->Partial ban, we should probably lift it for all source placements to avoid the API being confusing.  And I suspect there could be further implications to the cost planner etc.
* I don't know if we can find a way to do the numerically safe zero pad trick in a compiler-friendly way. I guess using torch.cond is a possibility, although it adds other complexity.  What about starting with a zeros tensor of the fixed/global shape on every rank, then using index_add to write the local shard data into the correct offset? This needs the offset to be a device-tensor though, i think

**Comment by pianpwk** [comment]
> X->Partial("sum") can be numerically dangerous if you do it by scaling, or can be SPMD-dangerous if you do it by padding (different ranks have different paddings, which means your graphs are non-spmd).
> 
> IIUC your PR is taking the latter approach, padding different sized chunks per rank as prefix/suffix.

fwiw, spmd_types did decide to accomodate the padding version, having a `varying_to_partial` method that takes rank as argument: https://github.com/meta-pytorch/spmd_types/blob/60705234a3ebb7a2ddd423cb3259f80985d270b2/spmd_types/_local.py#L980-L998

though it doesn't support uneven S yet

**Comment by wconstab** [comment]
Ok, if we are doing the padding approach for spmd_types we may as well do it here too. It's a separate question of being spmd-friendly for graphs. I wonder how much we care about that e.g for compile-on-one-rank compatibility- cc @aorenste 

I also wonder if it is inconsistent to prevent the planner from doing s->p if we claim it is a supported operation. I think we should at least have a clear story here.

**Comment by pianpwk** [comment]
> Ok, if we are doing the padding approach for spmd_types we may as well do it here too. It's a separate question of being spmd-friendly for graphs. I wonder how much we care about that e.g for compile-on-one-rank compatibility- cc @aorenste 

I vaguely remember one of the approaches to CooR spmd-ness was to hide non-spmd-ness behind custom ops, e.g. https://github.com/pytorch/pytorch/blob/9d4d32de638a913d0cc27885ee41e6e225afebda/torch/distributed/tensor/_collective_utils.py#L64-L79

@wconstab just wondering how that works for your concerns? I imagine it delegates more work to lowering, but could also be easier to catch?

**Comment by acisseJZhong** [comment]
@wconstab I applied your suggestion to change to the fixed/global -> index_add approach, would compiler compile same graph with this approach? 

> if it is inconsistent to prevent the planner from doing s->p if we claim it is a supported operation

I initially added s->p into the planner and it broke a bunch of tests as it changes the planner's decision? Therefore I banned it from the planner as I don't want the planner to do s->p(kind of meaningless?). Do you have a better idea in mind?

File: torch/distributed/tensor/placement_types.py [hunk 1]
```diff
@@ -636,6 +636,33 @@ def _replicate_to_shard(
...
+            logical_dim_size,
+            num_chunks,
+            mesh._sym_get_coordinate(mesh_dim),
+        )
+        torch._check(local_tensor.size(self.dim) == local_shard_size)
+        output_shape: list[IntLikeType] = list(local_tensor.shape)
+        output_shape[self.dim] = logical_dim_size
+        output = local_tensor.new_zeros(output_shape)
+        offset = torch.scalar_tensor(
+            local_offset, dtype=torch.int64, device=local_tensor.device
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:659]: Just checking, is local_offset a device tensor or a host scalar?  I still think structurally this index_add framing is cleaner as it produces structurally identical graphs, but I'm not sure if it literally solves the rank-identical graphs problem without validating whether we bake per-rank constants (Scalars) into the graph.  I'd be curious to know what happens, but I wouldn't block this PR over it.  cc @IvanKobzarev @aorenste

**Review by wconstab — COMMENTED** [review]
I think this is OK. But we should document our reasoning at least (the design, in code comment block), and make sure our API docs are clear too.

The important caveat is: “valid explicit redistribution” and “eligible implicit planner move” are different sets.  The reason we prefer this rather than having the sets be equal is partly due to a weakness in our cost model- it models comm cost only, not memory usage or computation (copy) time. If we put S->P as zero cost, we'd over-prefer it in the planner, and that would be worse than skipping it as an option.

For completeness sake, do we have an argument for supporting only explicit S->P rather than also R->P?

We should also check the reasoning between the transform-planner and the strategy-planner.
transform-planner implements S->P with zero cost now, while strategy planner has inf cost.  Is this inconsistency a problem? We could argue that as long as the transform planner DOES implement every finite cost transition in the strategy planner, we're good. A more extreme guarantee would be, they both follow the same paths always with the same costs.

**Comment by acisseJZhong** [comment]
Added doc string to clarify that every transition selected by the strategy planner must be executable by the transform planner, while the transform planner may support additional transitions used only by explicit redistribution. 

> For completeness sake, do we have an argument for supporting only explicit S->P rather than also R->P?

I assume user would never want to explicitly redistribute R->P? Also codex told me that `Replicate -> Partial("sum") divides by the mesh size and may introduce floating-point error, so keep this lossy conversion out of the public API.

File: torch/distributed/tensor/placement_types.py (hunk 1 above)
Comment (acisseJZhong) [reply torch/distributed/tensor/placement_types.py:659]: local_offset is host-side shape metadata represented as an int or SymInt

**Comment by tianyu-l** [comment]
> I assume user would never want to explicitly redistribute R->P? Also codex told me that `Replicate -> Partial("sum") divides by the mesh size and may introduce floating-point error, so keep this lossy conversion out of the public API.

The proper way probably should be only keeping value on rank 0, i.e. Owned-style Partial(sum).

With that I don't see the value of putting S->P as an intermediate step where either S is not original source or P is not final destination.

File: torch/distributed/tensor/_ops/single_dim_strategy.py [hunk 2]
```diff
@@ -943,6 +943,11 @@ def _get_neighbor_placements(
     - Shard(d1) -> Shard(d2): all-to-all, valid if neither d1 nor d2 sharded right
     - Partial -> Replicate: allreduce, always valid
     - Partial -> Shard(d): reduce-scatter, valid if d not sharded to the right
+
+    Shard -> Partial("sum") is intentionally absent. Although explicit
+    redistribution supports it, the conversion materializes a logical-shape
+    local tensor. The communication-only strategy cost model cannot represent
+    that memory and copy cost, so implicit planning must not select it.
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/single_dim_strategy.py]: nit: logical-shape -> global-shape 

must not select it -> we avoid selecting it based only on comm cost without considering other costs

**Review by wconstab — APPROVED** [review]
lgtm. thanks for making those changes!
